src/helpers/SparseCalldataSegmentLib.sol

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.25;
+
+/// @title Sparse Calldata Segment Library
+/// @notice Library for working with sparsely-packed calldata segments, identified with an index.
+/// @dev The first byte of each segment is the index of the segment.
+/// To prevent accidental stack-to-deep errors, the body and index of the segment are extracted separately, rather
+/// than inline as part of the tuple returned by `getNextSegment`.
+library SparseCalldataSegmentLib {
+    /// @notice Splits out a segment of calldata, sparsely-packed
+    /// @param source The calldata to extract the segment from
```

Comment:
Could be useful to document the expected format of the source, e.g., `[ len(segment0) - 1] [ segment0 ] ... [ len(segmentN) - 1 ][ segmentN ]`

Also maybe could be easier to reason about if the length of the segment included the index byte?

Reply:
I think there might have been a versioning issue - previously I didn't include the index in the segment length, but I've since updated it to include it. And sounds good on the comment, will add it in as a "spec" for the encoding.